exonum-java-binding-core/rust/integration_tests/tests/jni_cache.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate integration_tests;
+extern crate java_bindings;
+#[macro_use]
+extern crate lazy_static;
+
+use integration_tests::vm::create_vm_for_tests_with_fake_classes;
+use java_bindings::{jni::JavaVM, utils::jni_cache, JniExecutor, MainExecutor};
+
+use std::{
+    sync::{Arc, Barrier},
+    thread::spawn,
+};
+
+lazy_static! {
+    pub static ref VM: Arc<JavaVM> = create_vm_for_tests_with_fake_classes();
+    pub static ref EXECUTOR: MainExecutor = MainExecutor::new(VM.clone());
+}
+
+#[test]
+fn concurrent_cache_read() {
```

Comment:
I'd consider explicitly stating that this test is not supposed to reliably catch synchronization errors because that will require _a lot_ more iterations, but that, in turn, requires starting in the initial state (which in case of statics, will unnecessarily complicate the thing).